File: src/main/java/com/werfad/JumpHandler.java
package com.werfad;

import com.intellij.openapi.actionSystem.*;
import com.intellij.openapi.editor.Caret;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.actionSystem.EditorActionHandler;
import com.intellij.openapi.editor.actionSystem.EditorActionManager;
import com.intellij.openapi.editor.actionSystem.TypedAction;
import com.intellij.openapi.editor.actionSystem.TypedActionHandler;
import com.intellij.openapi.util.TextRange;
import com.werfad.finder.*;
import com.werfad.utils.EditorUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static com.intellij.openapi.actionSystem.IdeActions.ACTION_GOTO_DECLARATION;


public final class JumpHandler implements TypedActionHandler {
    private static final JumpHandler INSTANCE = new JumpHandler();

    static final int MODE_CHAR1 = 0;
    static final int MODE_CHAR2 = 1;
    static final int MODE_WORD0 = 2;
    static final int MODE_WORD1 = 3;
    static final int MODE_LINE = 4;
    static final int MODE_WORD1_DECLARATION = 5;

    private static TypedActionHandler mOldTypedHandler;
    private static EditorActionHandler mOldEscActionHandler;
    private static final MarksCanvas mMarksCanvas = new MarksCanvas();

    private static boolean isStart;

    private static Finder finder;
    private static Optional<Runnable> onJump = Optional.empty(); // Runnable that is called after jump

    private static List<MarksCanvas.Mark> lastMarks = Collections.emptyList();
    private static boolean isCanvasAdded = false;

    static JumpHandler getInstance() {
        return INSTANCE;
    }

    public void execute(@NotNull Editor e, char c, @NotNull DataContext dc) {
        List<MarksCanvas.Mark> marks = finder.input(e, c, lastMarks);
        if (marks != null) {
            lastMarks = marks;
            jumpOrShowCanvas(e, lastMarks);
        }
    }

    private final EditorActionHandler escActionHandler = new EditorActionHandler() {
        @Override
        protected void doExecute(@NotNull Editor editor, @Nullable Caret caret, DataContext dataContext) {
            stop();
        }
    };

    private void jumpOrShowCanvas(Editor e, List<MarksCanvas.Mark> marks) {
        if (marks.isEmpty()) {
            stop();
        } else if (marks.size() == 1) {
            // only one found, just jump to it
            e.getCaretModel().moveToOffset(marks.get(0).getOffset());
            stop();
            onJump.orElse(() -> {}).run();
        } else {
            if (!isCanvasAdded) {
                mMarksCanvas.sync(e);
                e.getContentComponent().add(mMarksCanvas);
                e.getContentComponent().repaint();
                isCanvasAdded = true;
            }

            mMarksCanvas.setData(marks);
        }

    }

    /**
     * start search mode
     *
     * @param mode mode enum, see {@link #MODE_CHAR1} {@link #MODE_CHAR2} etc
     */
    public final void start(int mode, AnActionEvent anActionEvent) {
        if (isStart) return;
        isStart = true;
        Editor editor = anActionEvent.getData(CommonDataKeys.EDITOR);
        if (editor == null) return;

        EditorActionManager manager = EditorActionManager.getInstance();
        TypedAction typedAction = manager.getTypedAction();

        mOldTypedHandler = typedAction.getRawHandler();
        typedAction.setupRawHandler(this);

        mOldEscActionHandler = manager.getActionHandler(IdeActions.ACTION_EDITOR_ESCAPE);
        manager.setActionHandler(IdeActions.ACTION_EDITOR_ESCAPE, escActionHandler);

        onJump = Optional.empty();

        switch (mode) {
            case MODE_CHAR1:
                finder = new Char1Finder();
                break;
            case MODE_CHAR2:
                finder = new Char2Finder();
                break;
            case MODE_WORD0:
                finder = new Word0Finder();
                break;
            case MODE_WORD1:
                finder = new Word1Finder();
                break;
            case MODE_LINE:
                finder = new LineFinder();
                break;
            case MODE_WORD1_DECLARATION:
                finder = new Word1Finder();
                onJump = Optional.of(() -> ActionManager
                        .getInstance()
                        .getAction(ACTION_GOTO_DECLARATION)
                        .actionPerformed(anActionEvent));
                break;
            default:
                throw new RuntimeException("Invalid start mode: " + mode);
        }

        TextRange visibleBorderOffset = EditorUtils.getVisibleRangeOffset(editor);
        String visibleString = editor.getDocument().getText(visibleBorderOffset);
        List<MarksCanvas.Mark> marks = finder.start(editor, visibleString, visibleBorderOffset);
        if (marks != null) {
            lastMarks = marks;
            this.jumpOrShowCanvas(editor, lastMarks);
        }
    }

    private void stop() {
        if (isStart) {
            isStart = false;
            EditorActionManager manager = EditorActionManager.getInstance();
            manager.getTypedAction().setupRawHandler(mOldTypedHandler);
            if (mOldEscActionHandler != null) {
                manager.setActionHandler(IdeActions.ACTION_EDITOR_ESCAPE, mOldEscActionHandler);
            }

            Container parent = mMarksCanvas.getParent();
            if (parent != null) {
                parent.remove(mMarksCanvas);
                parent.repaint();
            }

            isCanvasAdded = false;
        }
    }
}

File: src/main/kotlin/com/werfad/JumpHandler.kt
package com.werfad

import com.intellij.openapi.actionSystem.*
import com.intellij.openapi.command.CommandProcessor
import com.intellij.openapi.command.UndoConfirmationPolicy
import com.intellij.openapi.editor.Caret
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.actionSystem.DocCommandGroupId
import com.intellij.openapi.editor.actionSystem.EditorActionHandler
import com.intellij.openapi.editor.actionSystem.EditorActionManager
import com.intellij.openapi.editor.actionSystem.TypedAction
import com.intellij.openapi.editor.actionSystem.TypedActionHandler
import com.intellij.openapi.fileEditor.ex.IdeDocumentHistory
import com.werfad.finder.*
import com.werfad.utils.getVisibleRangeOffset

object JumpHandler : TypedActionHandler {
    private var mOldTypedHandler: TypedActionHandler? = null
    private var mOldEscActionHandler: EditorActionHandler? = null
    private val mMarksCanvas: MarksCanvas by lazy { MarksCanvas() }
    private var isStart = false
    private lateinit var finder: Finder
    private var onJump: (() -> Unit)? = null // Runnable that is called after jump
    private var lastMarks: List<MarksCanvas.Mark> = emptyList()
    private var isCanvasAdded = false

    override fun execute(e: Editor, c: Char, dc: DataContext) {
        val marks = finder.input(e, c, lastMarks)
        if (marks != null) {
            lastMarks = marks
            jumpOrShowCanvas(e, lastMarks)
        }
    }

    private val escActionHandler: EditorActionHandler = object : EditorActionHandler() {
        override fun doExecute(editor: Editor, caret: Caret?, dataContext: DataContext) {
            stop()
        }
    }

    private fun jumpOrShowCanvas(e: Editor, marks: List<MarksCanvas.Mark>) {
        when {
            marks.isEmpty() -> {
                stop()
            }
            marks.size == 1 -> {
                // only one found, just jump to it
                val caret = e.caretModel.currentCaret
                if (caret.hasSelection()) {
                    val downOffset =
                        if (caret.selectionStart == caret.offset)
                            caret.selectionEnd
                        else
                            caret.selectionStart
                    caret.setSelection(downOffset, marks[0].offset)
                }
                // Shamelessly robbed from AceJump: https://github.com/acejump/AceJump/blob/99e0a5/src/main/kotlin/org/acejump/action/TagJumper.kt#L87
                with(e) {
                    project?.let { project ->
                        CommandProcessor.getInstance().executeCommand(
                            project, {
                                with(IdeDocumentHistory.getInstance(project)) {
                                    setCurrentCommandHasMoves()
                                    includeCurrentCommandAsNavigation()
                                    includeCurrentPlaceAsChangePlace()
                                }
                            }, "KJumpHistoryAppender", DocCommandGroupId.noneGroupId(document),
                            UndoConfirmationPolicy.DO_NOT_REQUEST_CONFIRMATION, document
                        )
                    }
                }
                caret.moveToOffset(marks[0].offset)

                stop()
                onJump?.invoke()
            }
            else -> {
                if (!isCanvasAdded) {
                    mMarksCanvas.sync(e)
                    e.contentComponent.add(mMarksCanvas)
                    e.contentComponent.repaint()
                    isCanvasAdded = true
                }
                mMarksCanvas.setData(marks)
            }
        }
    }

    /**
     * start search mode
     *
     * @param mode mode enum, see [.MODE_CHAR1] [.MODE_CHAR2] etc
     */
    fun start(mode: JumpMode, anActionEvent: AnActionEvent) {
        if (isStart) return
        isStart = true
        val editor = anActionEvent.getData(CommonDataKeys.EDITOR) ?: return
        val manager = EditorActionManager.getInstance()
        val typedAction = TypedAction.getInstance()
        mOldTypedHandler = typedAction.rawHandler
        typedAction.setupRawHandler(this)
        mOldEscActionHandler = manager.getActionHandler(IdeActions.ACTION_EDITOR_ESCAPE)
        manager.setActionHandler(IdeActions.ACTION_EDITOR_ESCAPE, escActionHandler)
        onJump = null
        when (mode) {
            JumpMode.Char1 -> finder = Char1Finder()
            JumpMode.Char2 -> finder = Char2Finder()
            JumpMode.Word0 -> finder = Word0Finder()
            JumpMode.Word1 -> finder = Word1Finder()
            JumpMode.Line -> finder = LineFinder()
            JumpMode.Word1Declaration -> {
                finder = Word1Finder()
                onJump = {
                    ActionManager
                        .getInstance()
                        .getAction(IdeActions.ACTION_GOTO_DECLARATION)
                        .actionPerformed(anActionEvent)
                }
            }
        }
        val visibleBorderOffset = editor.getVisibleRangeOffset()
        val visibleString = editor.document.getText(visibleBorderOffset)
        val marks = finder.start(editor, visibleString, visibleBorderOffset)
        if (marks != null) {
            lastMarks = marks
            jumpOrShowCanvas(editor, lastMarks)
        }
    }

    private fun stop() {
        if (isStart) {
            isStart = false
            val manager = EditorActionManager.getInstance()
            TypedAction.getInstance().setupRawHandler(mOldTypedHandler!!)
            if (mOldEscActionHandler != null) {
                manager.setActionHandler(IdeActions.ACTION_EDITOR_ESCAPE, mOldEscActionHandler!!)
            }
            val parent = mMarksCanvas.parent
            if (parent != null) {
                parent.remove(mMarksCanvas)
                parent.repaint()
            }
            isCanvasAdded = false
        }
    }
}